modules/apigee_edge_teams/src/TeamContextManager.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge_teams;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Url;
+
+/**
+ * Describes the `apigee_edge_teams.context_manager` service.
+ *
+ * This service is responsible for understanding the context of the current
+ * route. Context will either be developer context or team context.
+ */
+class TeamContextManager implements TeamContextManagerInterface {
```

Comment:
The only thing that does is forces a decorator pattern over inheritance. Why is it important to do that?

Reply:
1. https://matthiasnoback.nl/2018/09/final-classes-by-default-why/
2. https://matthiasnoback.nl/2018/08/when-to-add-an-interface-to-a-class/
3. https://codeburst.io/inheritance-is-evil-stop-using-it-6c4f1caf5117

How this all applies to Symfony and Drupal 8 service classes?

You create an interface for a service in Drupal 8/Symfony because you and external developers should (must) depend on only public methods exposed on a service interface. Which means, everybody should only depend on the expected behavior of service methods instead of their concrete implementation (that you may ship in a class with your module).

Thanks for this approach, external developers can easily replace completely or just decorate the service implementations (classes) that you ship with your module with their own, they only need to create something which is compatible with the service interface.

Services usually have external dependencies that they get in constructor parameters. Drupal 8 BC policy does states that [constructor parameters may change between _minor_ versions](https://www.drupal.org/core/d8-bc-policy#constructors). _(Not sure how many Drupal developers knows what they should extend/override and what they should based on the BC policy. Developers do not read doc usually...)_

What does it mean for external developers?
If they still prefer inheritance over encapsulation then their code gets broken when a service's constructor gets changed.

How can we help them?
This is easy, we mark service classes as final so they do not try to extend it. They do not need to worry about constructor parameter changes because they cannot extend our concrete service implementations. Although they can always [inject the original service with "inner"](https://www.blueoakinteractive.com/blog/service-decorators-drupal-8) and leverage the actually re-usable components from the service - the public methods.

Just think about it, this is not just about constructors, the same problem occurs for public and protected methods and properties. As a maintainer, you are going to have bad days if you try to change these on your non-final classes because you may introduce BC breaking changes with those changes.